### ISSUE TITLE
Server shutdown while connection is still open

### ISSUE THREAD

**Issue by heindl-solutions** [oatpp/oatpp#476]
We have set up the server similar to
https://github.com/oatpp/example-server-stop/blob/master/src/App_StopByConditionCheck.cpp

If a client opens a connection to `localhost:8000`, the function `HttpProcessor::Task::run()` is entered 3 times, and left immediately 2 times.
This means that one thread handling the open connection is kept open, as might be expected.

But now there is a problem if we want to shut down the server, because the loop (HttpProcessor.cpp, l. 237)
```
 do {
      connectionState = HttpProcessor::processNextRequest(resources);
    } while (connectionState == ConnectionState::ALIVE);
```

is still running until the client closes the connection, although we have stopped and deleted the underlying resources already!

Thus it seems there is currently no way to shut down the server in a clean way.

We are using Oatpp 1.2.5 on x64-win64.

**Comment by bamkrs** [comment]
Hi Andreas, 
yes, currently the stopping of the server is only defined for new connections. This way, already established connections can be handled and gracefully closed, thus preventing dangling I.E. Database calls or file writes (if correctly implemented/synchronised on the user side). Oat++ was never designed to be shut down in the first place so all those functions are gradually added. 

As of version 1.3.0 (upcoming release) for example, we at least count the created connections and keep everything alive until the last connection was served (`stop()` actually waits/blocks for connections to be closed!). Thus, we track everything inside Oat++ and the user does not need to anymore. 
Maybe give the 1.3.0 branch a try and see if it fixes your problem.

**Comment by sgallou** [comment]
I encoutered the same problem, which makes my server crash (under Windows) after processing a request containing a `Connection=keep-alive` header. 

The workaround I found is to use my own connection provider, which is keeping trace of created connections, to be able to close they (= free sockets) :

```cpp
class MyConnectionProvider final : public oatpp::network::tcp::server::ConnectionProvider
{
public:
   MyConnectionProvider(const oatpp::network::Address& address,
      bool useExtendedConnections = false)
      : ConnectionProvider(address,
         useExtendedConnections)
   {
   }

   ~MyConnectionProvider() override
   {
      for (const auto& c : m_connections)
         c->close();
   }

   static std::shared_ptr<MyConnectionProvider> createShared(const oatpp::network::Address& address,
      bool useExtendedConnections = false)
   {
      return std::make_shared<MyConnectionProvider>(address,
         useExtendedConnections);
   }

   std::shared_ptr<oatpp::data::stream::IOStream> get() override
   {
      auto connection = ConnectionProvider::get();
      if (connection)
         m_connections.emplace(std::static_pointer_cast<oatpp::network::tcp::Connection>(connection));
      return connection;
   }

private:
   std::set<std::shared_ptr<oatpp::network::tcp::Connection>> m_connections;
};
```

**Comment by lganzzzo** [comment]
Hey @sgallou ,

This is a possible solution - however, in your case, you keep all Connection objects alive even those that are already closed.
Since you are using `std::shared_ptr` to keep them.

What you want to add - is to remove connections from the provider's list when the connection is destructed. And of cause change `m_connections` list to store `weak_ptr` instead


*We are planning to fix this issue in oatpp `1.3.0`*

**Comment by sgallou** [comment]
Yes you're right, I will fix this, while waiting for 1.3.0.
Thanks for your attention !

**Comment by sgallou** [comment]
> We are planning to fix this issue in oatpp `1.3.0`

Hi @lganzzzo, when do you plan to publish this `1.3.0` ?

**Comment by lganzzzo** [comment]
Fixed in `1.3.0`. See [proper-server-stoppage](https://github.com/oatpp/oatpp/blob/master/changelog/1.3.0.md#proper-server-stoppage)
